### PR TITLE
[llvm-exegesis] Add support for loading X86 segment registers

### LINKED ISSUE #76340: [llvm-exegesis] Cannot execute snippets using segment registers
Currently, it is difficult to execute snippets on x86 that use the segment registers like `%fs` or `%gs` as there is no way currently to control their values. So snippets like the following:
```asm
# LLVM-EXEGESIS-DEFREG R12 12345600
# LLVM-EXEGESIS-MEM-DEF MEM 4096 0000000012345600
# LLVM-EXEGESIS-MEM-MAP MEM 23456242130944
        movq    %fs:16, %r12
```
Just trying to set the segment register value with a `LLVM-EXEGESIS-DEFREG` annotation doesn't work as the standard move instructions can't move values into the segment registers. There are some user space instructions introduced in sandy bridge and zen that enable setting it in userspace, but it can also be done with a system call (on Linux).

This only makes sense in the subprocess mode as segment registers are only really useful when accessing memory.

### PR DESCRIPTION
**Pull request by boomanaiden154**
This patch adds support for setting the X86 segment registers. These registers are used in quite a few basic blocks in BHive and similar datasets, so being able to set them is necessary to ensure consistent runs as the live-in values of fs and gs can change across runs.

Fixes #76340.

### CODE REVIEW

**Comment by boomanaiden154** [comment]
Fixes #76368.

**Comment by boomanaiden154** [comment]
Fixes #76340.

File: llvm/tools/llvm-exegesis/lib/X86/Target.cpp [hunk 1]
```diff
@@ -907,9 +908,62 @@ void ExegesisX86Target::decrementLoopCounterAndJump(
...
+  // Preserve RCX and R11 (clobbered by the system call), and RAX, RDI, and RSI
+  // (used to make the system call). Preserve the registers here as we don't
+  // want to make any assumptions about the ordering of what registers are
+  // loaded in first, and we might have already loaded in registers that we are
+  // going to be clobbering here.
+  generateRegisterStackPush(X86::RAX, loadSegmentRegisterCode);
+  generateRegisterStackPush(X86::RDI, loadSegmentRegisterCode);
+  generateRegisterStackPush(X86::RSI, loadSegmentRegisterCode);
+  generateRegisterStackPush(X86::RCX, loadSegmentRegisterCode);
+  generateRegisterStackPush(X86::R11, loadSegmentRegisterCode);
```
Comment (legrosbuffle) [inline llvm/tools/llvm-exegesis/lib/X86/Target.cpp:940]: Please refactor this (this code is already present on line 1174)

File: llvm/tools/llvm-exegesis/lib/X86/Target.cpp [hunk 2]
```diff
@@ -907,9 +908,62 @@ void ExegesisX86Target::decrementLoopCounterAndJump(
...
+  // Generate the instructions to make the arch_prctl system call to set
+  // the registers.
+  assert(Reg == X86::FS ||
+         Reg == X86::GS &&
+             "Only the segment registers GS and FS are supported");
+  int SyscallCode = 0;
+  SyscallCode = Reg == X86::FS ? SyscallCode | ARCH_SET_FS : SyscallCode;
```
Comment (legrosbuffle) [inline llvm/tools/llvm-exegesis/lib/X86/Target.cpp:947]: This is quite cryptic. Why not:

```
if (Reg == X86::FS) {
  SyscallCode = ARCH_SET_FS;
} else if (Reg == X86::GS) {
  SyscallCode = ARCH_SET_GS;
} else {
   llvm_unreacahble(...);
}
```

File: llvm/tools/llvm-exegesis/lib/X86/Target.cpp [hunk 3]
```diff
@@ -924,38 +924,61 @@ void generateSyscall(long SyscallNumber, std::vector<MCInst> &GeneratedCode) {
   GeneratedCode.push_back(MCInstBuilder(X86::SYSCALL));
 }
 
+constexpr std::array<unsigned, 6> SyscallArgumentRegisters{
+    X86::RDI, X86::RSI, X86::RDX, X86::R10, X86::R8, X86::R9};
+
+static void saveSyscallRegisters(std::vector<MCInst> &GeneratedCode,
+                                 unsigned ArgumentCount) {
+  assert(ArgumentCount < 6 &&
```
Comment (legrosbuffle) [inline llvm/tools/llvm-exegesis/lib/X86/Target.cpp:932]: `<=` ?

File: llvm/tools/llvm-exegesis/lib/X86/Target.cpp [hunk 4]
```diff
@@ -924,38 +924,61 @@ void generateSyscall(long SyscallNumber, std::vector<MCInst> &GeneratedCode) {
...
+  // Preserve RAX (used for the syscall number/return value).
+  generateRegisterStackPush(X86::RAX, GeneratedCode);
+  // Preserve the registers used to pass arguments to the system call.
+  for (unsigned I = 0; I < ArgumentCount; ++I)
+    generateRegisterStackPush(SyscallArgumentRegisters[I], GeneratedCode);
+}
+
+static void restoreSyscallRegisters(std::vector<MCInst> &GeneratedCode,
+                                    unsigned ArgumentCount) {
+  assert(ArgumentCount < 6 &&
```
Comment (legrosbuffle) [inline llvm/tools/llvm-exegesis/lib/X86/Target.cpp:946]: ditto

**Review by legrosbuffle — APPROVED** [review]

**Comment by nico** [comment]
The reland breaks building on windows: http://45.33.8.238/

**Comment by boomanaiden154** [comment]
> The reland breaks building on windows: http://45.33.8.238/

3e6e09609d8cef3c7cc99ced1a043869c8e984b8 will hopefully fix that. I'll keep an eye on your CI and the buildbots. If that doesn't fix it, I'll pull the patch and retest on a couple other platforms before relanding it again to make sure I have the preprocessor setup correct.

**Comment by nico** [comment]
Thanks, that helped!

Mind spelling `and` as `&&` to be consistent with the rest of the codebase though?

**Comment by boomanaiden154** [comment]
> Thanks, that helped!
> 
> Mind spelling `and` as `&&` to be consistent with the rest of the codebase though?

Fixed in 589a24b0b75d2def0b6ebf64953d367c725102d3. Meant to do that before committing the fix but forgot about it. Thanks for the suggestion!

**Pull request merged**